invenio_rdm_records/services/services.py

```diff
@@ -131,3 +148,229 @@ def oai_result_item(self, identity, oai_record_source):
+                        "in order to submit the access request: %(url)s"
+                    ),
+                    url=verify_url,
+                ),
+            )
+        )
+
+        return {
+            "message": _("Verification link sent out, please check your e-mail inbox"),
+        }
+
+    @unit_of_work()
+    def create_guest_access_request(self, identity, token, expand=False, uow=None):
+        """Use the access request token to create an access request."""
+        # TODO proper permission checks
+        if authenticated_user in identity.provides:
+            raise PermissionDeniedError("request_guest_access")
+
+        access_token = AccessRequestToken.get_by_token(token)
+        if access_token is None:
+            return None
+
+        access_token_data = access_token.to_dict()
+        record = self.record_cls.pid.resolve(access_token_data["record_pid"])
+
+        # TODO try to find a duplicate request
+        req_cls = current_requests_service.record_cls
+        model_cls = req_cls.model_cls
+        requests = [
+            req_cls(rm.data, model=rm)
+            for rm in model_cls.query.filter(
+                model_cls.json["created_by"] == {"email": access_token.email},
+                model_cls.json["topic"] == {"record": access_token.record_pid},
+            )
+            if rm.data and rm.data["type"] == GuestAccessRequest.type_id
+        ]
+
+        if any((req.is_open for req in requests)):
+            # TODO DuplicateRequestError
+            raise PermissionDeniedError()
+
+
+        data = {
+            "payload": {
+                "permission": record.parent.access.settings.default_access_permission,
+                "email": access_token_data["email"],
+                "full_name": access_token_data["full_name"],
+                "token": access_token_data["token"],
+                "message": access_token_data.get("message") or "No comment",
+            }
+        }
+
+        receiver = None
+        record_owner = record.parent.access.owner.resolve()
+        if record_owner:
+            receiver = record_owner
+
+        if receiver is None:
+            pass
+
+        access_token.delete()
+        request = current_requests_service.create(
+            system_identity,  # TODO change
+            data,
+            GuestAccessRequest,
+            receiver,
+            creator=data["payload"]["email"],
+            topic=record,
+            expires_at=None,  # TODO expiration date
+            expand=expand,
+            uow=uow,
+        )
+
+        if request.errors:
+            # TODO maybe this should be returned as payload of an exception?
+            return request
+
+        message = {
+            "payload": {
+                "content": data["payload"].get("message") or "No comment",
+            }
+        }
+
+        return current_requests_service.execute_action(
+            system_identity,  # TODO change
```

Comment:
TODO to be addressed?

Reply:
Yeah, this needs to be addressed as part of the issue regarding the guest's comments on the request.
In a nutshell, there's currently no good way to tell the services that _"the guest"_ (maybe with _"that email address"_) created the comment.
So for now, we're evading this by letting "the system" create the first comment on the request.